primitives/manta/src/xcm.rs

```diff
@@ -417,6 +417,69 @@ where
     }
 }
 
+use xcm::latest::{Instruction::*, Weight};
+
+/// Allows execution from `origin` if it is contained in `T` (i.e. `T::Contains(origin)`) taking
+/// payments into account.
+///
+/// Only allows for `TeleportAsset`, `WithdrawAsset`, `ClaimAsset` and `ReserveAssetDeposit` XCMs
+/// because they are the only ones that place assets in the Holding Register to pay for execution.
+pub struct AllowTopLevelPaidExecutionFrom<T>(PhantomData<T>);
+impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionFrom<T> {
+    fn should_execute<RuntimeCall>(
+        origin: &MultiLocation,
+        message: &mut Xcm<RuntimeCall>,
+        max_weight: Weight,
+        _weight_credit: &mut Weight,
+    ) -> Result<(), ()> {
+        log::trace!(
+            target: "xcm::barriers",
+            "AllowTopLevelPaidExecutionFrom origin: {:?}, message: {:?}, max_weight: {:?}, weight_credit: {:?}",
+            origin, message, max_weight, _weight_credit,
+        );
+        ensure!(T::contains(origin), ());
+        let mut iter = message.0.iter_mut();
+        let i = iter.next().ok_or(())?;
+        match i {
+            ReceiveTeleportedAsset(..)
+            | WithdrawAsset(..)
+            | ReserveAssetDeposited(..)
+            | ClaimAsset { .. } => (),
+            _ => return Err(()),
+        }
+        let mut i = iter.next().ok_or(())?;
+        while let ClearOrigin = i {
```

Comment:
This forces that there must be only ClearOrigin before any other instructions are executed.
So it force drops the origin of this message from a parachain sovereign account to nothing, in order to reduce the privilege of the message.
Don't think it can be combined in that match.